`datalake`: relax `translation_stm::max_collectible_offset()` value (and add `compaction_test.py`)

WillemKauf opened the pull request:
Previously, the `datalake::translation::translation_stm` would return its max collectible as the following:

https://github.com/redpanda-data/redpanda/blob/925707c3fa2f899dede79954060c67227fcf65d4/src/v/datalake/translation/state_machine.cc#L112-L122

This offset translation leads to an overly restrictive condition for the max collectible offset, due to the fact that it is translation batch unaware.

Here, the utility function `highest_log_offset_below_next()` is added, which returns the "equivalent" translated log offset for a given kafka offset, taking into account translation batches (which don't need to be translated, and thus shouldn't restrict the max collectible offset).

`translation_stm::max_collectible_offset()` now uses this function to relax its returned offset.

Additionally, a new test for compaction with an Iceberg enabled topic is added to `datalake/compaction_test.py`, with some enhancements to the `datalake_verifier` service to make it compaction aware.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Improvements

* Fixes an overly restrictive condition for retention in Iceberg-enabled topics.

WillemKauf committed:
`rptest`: make `datalake_verifier` compaction aware

By handling gaps in offsets and recording seen keys, we can validate the
correctness of a compacted log that has been translated (fully) into an
iceberg table.

WillemKauf committed:
`datalake`: add `compaction_test.py`

Adds a new `test_compaction` test, which uses the `KgoVerifierSeqConsumer`
to validate a fully compacted log, along with the `datalake_verifier`
service to validate the Iceberg table.

Also moves the contents of `compaction_gaps_test.py` into
`compaction_test.py`.

WillemKauf commented on tests/rptest/tests/datalake/datalake_verifier.py:
```diff
@@ -24,22 +24,24 @@
     """
-     Verifier that does the verification of the data in the redpanda Iceberg table. 
-     The verifier consumes offsets from specified topic and verifies it the data 
+     Verifier that does the verification of the data in the redpanda Iceberg table.
```
Trailing whitespace removal

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59935-->
<!--5efed726-7ec3-4830-bdd0-21a3166e09ab-->
<!--bd03b5f9-742c-44d7-9397-6f83b3cf291c-->
<!--c4c2db0a-2952-4993-b214-eace140d7106-->
<!--5f9eae59-faf0-4b3c-b81c-1f3127c10191-->
<!--fb9ed56f-8c44-42ec-a398-5423bee18981-->
<!--2140282f-1135-44f2-b6ea-c4541e9eedc6-->
<!--ef61063f-cb0c-4f21-914a-f909974cc34d-->
<!--fdb86bbd-ad2d-4db9-a669-72a73e97c4cb-->
<!--title-->
## Retry command for [Build#59935](https://buildkite.com/redpanda/redpanda/builds/59935)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_ChunkedRead == True)"}}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[2,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_TxRangeMaterialized == True, SpilloverManifestUploaded == True)"}}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_TxRangeMaterialized == True, SpilloverManifestUploaded == True)"}}
tests/rptest/tests/write_caching_fi_test.py::WriteCachingFailureInjectionTest.test_crash_all
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[2,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_TxRangeMaterialized == True)"}}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"virtual_host"],"test_case":{"name":"(TS_Read == True, SpilloverManifestUploaded == True)"}}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_TxRangeMaterialized == True)"}}
tests/rptest/tests/write_caching_fi_e2e_test.py::WriteCachingFailureInjectionE2ETest.test_crash_all@{"use_transactions":false}
tests/rptest/tests/datalake/datalake_e2e_test.py::DatalakeE2ETests.test_topic_lifecycle@{"cloud_storage_type":1,"filesystem_catalog_mode":false}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_Timequery == True)"}}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"path"],"test_case":{"name":"(TS_Read == True, TS_TxRangeMaterialized == True)"}}
tests/rptest/tests/datalake/datalake_e2e_test.py::DatalakeE2ETests.test_topic_lifecycle@{"cloud_storage_type":1,"filesystem_catalog_mode":true}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_Timequery == True, SpilloverManifestUploaded == True)"}}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"path"],"test_case":{"name":"(TS_Read == True, TS_TxRangeMaterialized == True, SpilloverManifestUploaded == True)"}}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"virtual_host"],"test_case":{"name":"(TS_Read == True, AdjacentSegmentMergerReupload == True)"}}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--14821559-520c-4b6c-9797-13c760d0c00a-->
<!--TEST RESULTS-->
<!--e0163da0-c33a-4125-83bc-c3951b710531-->
<!--TEST RESULTS-->
<!--eba6de1e-cde5-4b47-bad7-7425094ca9b4-->
<!--TEST RESULTS-->
<!--5c6b7f8a-e61d-41ce-bcb9-c4beaa700e31-->
<!--TEST RESULTS-->
<!--4cf50fd7-d070-422b-8971-21704bb64ef3-->
<!--TEST RESULTS-->
<!--3d4ac5eb-ac95-40e9-8ef6-fb61b4165183-->
<!--TEST RESULTS-->
<!--2cc7ce1c-0b52-4d68-9f19-be7939799a1d-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59935">build#59935</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.filesystem_catalog_mode=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc18-c804-449e-983b-f6040b48eed2 | FAIL | 0/6 |
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.filesystem_catalog_mode=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc1c-24f3-40db-9e3c-b72686edcfd1 | FAIL | 0/6 |
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.filesystem_catalog_mode=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc18-c806-4851-8e40-029c7bdf36d7 | FAIL | 0/6 |
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.filesystem_catalog_mode=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc1c-24f4-48f5-b848-e36477ea95e1 | FAIL | 0/6 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc18-c803-4c2f-9cd0-86f9a8dd5064 | FLAKY | 5/6 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.ABS.2.virtual_host.test_case=.TS_Read==True.TS_TxRangeMaterialized==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc18-c807-4275-8982-8723111a2347 | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.ABS.2.virtual_host.test_case=.TS_Read==True.TS_TxRangeMaterialized==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc1c-24f4-48f5-b848-e36477ea95e1 | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.ABS.2.virtual_host.test_case=.TS_Read==True.TS_TxRangeMaterialized==True.SpilloverManifestUploaded==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc18-c803-4c2f-9cd0-86f9a8dd5064 | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.ABS.2.virtual_host.test_case=.TS_Read==True.TS_TxRangeMaterialized==True.SpilloverManifestUploaded==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc1c-24f6-49d5-ae5a-d5eb3497ad6e | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.path.test_case=.TS_Read==True.TS_TxRangeMaterialized==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc18-c804-449e-983b-f6040b48eed2 | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.path.test_case=.TS_Read==True.TS_TxRangeMaterialized==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc1c-24f2-467e-a057-0e4a790311ae | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.path.test_case=.TS_Read==True.TS_TxRangeMaterialized==True.SpilloverManifestUploaded==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc18-c806-4851-8e40-029c7bdf36d7 | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.path.test_case=.TS_Read==True.TS_TxRangeMaterialized==True.SpilloverManifestUploaded==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc1c-24f3-40db-9e3c-b72686edcfd1 | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.virtual_host.test_case=.TS_Read==True.AdjacentSegmentMergerReupload==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc1c-24f6-49d5-ae5a-d5eb3497ad6e | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.virtual_host.test_case=.TS_Read==True.SpilloverManifestUploaded==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc18-c807-4275-8982-8723111a2347 | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.virtual_host.test_case=.TS_Read==True.SpilloverManifestUploaded==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc1c-24f4-48f5-b848-e36477ea95e1 | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.virtual_host.test_case=.TS_Read==True.TS_ChunkedRead==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc18-c803-4c2f-9cd0-86f9a8dd5064 | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.virtual_host.test_case=.TS_Read==True.TS_Timequery==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc18-c804-449e-983b-f6040b48eed2 | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.virtual_host.test_case=.TS_Read==True.TS_Timequery==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc1c-24f2-467e-a057-0e4a790311ae | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.virtual_host.test_case=.TS_Read==True.TS_Timequery==True.SpilloverManifestUploaded==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc18-c806-4851-8e40-029c7bdf36d7 | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.virtual_host.test_case=.TS_Read==True.TS_Timequery==True.SpilloverManifestUploaded==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc1c-24f3-40db-9e3c-b72686edcfd1 | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.virtual_host.test_case=.TS_Read==True.TS_TxRangeMaterialized==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc18-c807-4275-8982-8723111a2347 | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.virtual_host.test_case=.TS_Read==True.TS_TxRangeMaterialized==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc1c-24f4-48f5-b848-e36477ea95e1 | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.virtual_host.test_case=.TS_Read==True.TS_TxRangeMaterialized==True.SpilloverManifestUploaded==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc18-c803-4c2f-9cd0-86f9a8dd5064 | FAIL | 0/1 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.virtual_host.test_case=.TS_Read==True.TS_TxRangeMaterialized==True.SpilloverManifestUploaded==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc1c-24f6-49d5-ae5a-d5eb3497ad6e | FAIL | 0/1 |
| rptest.tests.write_caching_fi_e2e_test.WriteCachingFailureInjectionE2ETest.test_crash_all.use_transactions=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc18-c804-449e-983b-f6040b48eed2 | FAIL | 0/1 |
| rptest.tests.write_caching_fi_e2e_test.WriteCachingFailureInjectionE2ETest.test_crash_all.use_transactions=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc1c-24f4-48f5-b848-e36477ea95e1 | FAIL | 0/1 |
| rptest.tests.write_caching_fi_test.WriteCachingFailureInjectionTest.test_crash_all | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc18-c803-4c2f-9cd0-86f9a8dd5064 | FAIL | 0/1 |
| rptest.tests.write_caching_fi_test.WriteCachingFailureInjectionTest.test_crash_all | ducktape | https://buildkite.com/redpanda/redpanda/builds/59935#0193dc1c-24f3-40db-9e3c-b72686edcfd1 | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60023">build#60023</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60023#0193e62a-cfd5-4342-9d5c-bd82d737eba0 | FLAKY | 2/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60079">build#60079</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_concurrent_truncations.cloud_storage_enabled=True.truncate_point=start_offset | ducktape | https://buildkite.com/redpanda/redpanda/builds/60079#0193f494-6494-4f55-ac0e-69cf1956752e | FLAKY | 5/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60368">build#60368</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/60368#01944294-d5f0-4c98-8132-4655bda7db82 | FLAKY | 5/6 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60368#019442af-789d-4ffa-a231-a3a2c2539ba9 | FLAKY | 1/6 |
| rptest.transactions.stream_verifier_test.StreamVerifierTest.test_simple_produce_consume_txn_with_add_node | ducktape | https://buildkite.com/redpanda/redpanda/builds/60368#01944294-d5ed-4718-a54f-7ac9bb15fdcc | FLAKY | 5/6 |
| storage_e2e_single_thread_rpunit.storage_e2e_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60368#01944252-4124-4c76-85c2-f3165a0ba962 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60544">build#60544</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rm_stm_tests_rpunit.rm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60544#01944d38-a71c-4bb4-86a4-b75b1322f1b0 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60563">build#60563</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rm_stm_tests_rpunit.rm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60563#01944e3e-7668-4063-8ce8-c9bff6ddcdc7 | FLAKY | 1/2 |
| rptest.tests.datalake.simple_connect_test.RedpandaConnectIcebergTest.test_translating_avro_serialized_records.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60563#01944e84-4faa-45d6-848a-06d1d9cef224 | FLAKY | 5/6 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60563#01944e84-4faa-45d6-848a-06d1d9cef224 | FLAKY | 1/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60597">build#60597</a></summary>


| test_id                                                                                               | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rm_stm_tests_rpunit.rm_stm_tests_rpunit                                                               | unit        | https://buildkite.com/redpanda/redpanda/builds/60597#019450cc-9bd3-4e5f-afb0-e1ec69fbb762 | FLAKY         | 1/2      |
| rm_stm_tests_rpunit.rm_stm_tests_rpunit                                                               | unit        | https://buildkite.com/redpanda/redpanda/builds/60597#019450cc-9cc5-4fda-8817-501bafa1c3fe | FLAKY         | 1/2      |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60597#01945126-d4c5-4f13-bf17-1e1af69bd8cc | FLAKY         | 5/6      |
</details>
<!--/body-->

WillemKauf commented:
Lot of `KgoVerifierProducer` failures, `panic: Out of order offset 0 (vs 0 20000)`.

Not sure if this is another `KgoVerifierProducer` issue or if something else has been broken.

The only related change I can see in `KgoVerifier` [was this](https://github.com/redpanda-data/kgo-verifier/pull/60/files#diff-819857411ab4917688be36b43e038c04def7cc8f374ce20aca11bccae2511628L377-L380), in which `pw.validOffsets.Insert()` is now called [under a lock in new function `OnAcked`](https://github.com/redpanda-data/kgo-verifier/pull/60/files#diff-819857411ab4917688be36b43e038c04def7cc8f374ce20aca11bccae2511628R206-R220) (but CI must have ran for this change many times before seeing these failures, so I am uncertain)


EDIT: Probably just because of the `oneshot()` changes I made. Reverted.

WillemKauf commented:
Force push to:
* Revert changes to `kgo_verifier_service::oneshot()`.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60016-->
<!--53006561-086e-41ae-9b22-4389ada15ee2-->
<!--title-->
## Retry command for [Build#60016](https://buildkite.com/redpanda/redpanda/builds/60016)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/compaction_test.py::CompactionTest.test_compaction@{"cloud_storage_type":1}
```
<!--/body-->

WillemKauf commented:
Force push to:
* Change compaction wait condition in `compaction_test.py`. Translation seems to slow the compaction process down quite a bit.

mmaslankaprv commented on src/v/datalake/translation/tests/translated_log_offset_test.cc:
```diff
@@ -0,0 +1,150 @@
+        b.add_batch(std::move(batch)).get();
+    };
+
+    make_and_add_record(model::record_batch_type::raft_data);
```
i think it would make sense to add a test case where we start from configuration batch as this is what it looks like in real life, configuration is always the first batch

WillemKauf replied:
Thanks for the callout here, I have added two new test cases which begin with configuration batches to `translated_log_offset_test`.

WillemKauf commented:
Force push to:
* Add two new tests to `translated_log_offset_test.cc`
* Remove early return in `get_translated_log_offset()` to correct behavior for the edge case of `kafka::offset{}`
* Add comment to `get_translated_log_offset()` declaration about its use.

bharathv commented on src/v/datalake/translation/utils.cc:
```diff
@@ -0,0 +1,24 @@
+namespace datalake::translation {
+
+model::offset
+get_translated_log_offset(ss::shared_ptr<storage::log> log, kafka::offset o) {
```
A little confused with the semantics here, consider the following example

```
C = config batch
K = kafka data batch

log ----------- [C, C, C, K, K, K, C, C, K]
log_offsets-----[0, 1, 2, 3, 4, 5, 6, 7, 8]
kaf_offsets-----[-, -, -, 0, 1, 2, -, -, 3]
```
get_translated_log_offset(log, kafka::offset{2})
nko = 3
log_offset_for_nko = 8
return 7

is that expected?

WillemKauf replied:
Yes, those are the expected semantics (I tried to draw this visual out in the added unit tests).

In your example, if you considered kafka::offset{2}, you can make the claim that all log offsets up to model::offset{7} have been translated. This is not inclusive of model::offset{8}, because that would require kafka::offset{3} having been translated- or rather, get_log_translated_offset(kafka::offset{3}) is expected to be model::offset{8}. 

Do the semantics make sense to you?

bharathv replied:
Ahhh, I got confused by what "translated" means in this (get_translated_log_offset) context .. I thought it was offset translation and not datalake translation. Let me take a look at the PR again.

bharathv commented on src/v/datalake/translation/state_machine.h:
```diff
@@ -44,6 +44,7 @@ class translation_stm final : public raft::persisted_stm<> {
 
     ss::future<std::error_code> reset_highest_translated_offset(
       kafka::offset new_translated_offset,
+      model::offset new_log_translated_offset,
```
nit: new_translated_log_offset?

WillemKauf replied:
Fixed

bharathv commented on tests/rptest/tests/datalake/datalake_verifier.py:
```diff
@@ -70,6 +72,8 @@ def __init__(self, redpanda: RedpandaService, topic: str,
         # map of last queried offset for each partition
         self._max_queried_offsets = {}
         self._last_checkpoint = {}
+        self._expected_compacted_keys = set()
```
nit: could use a comment (more generally how the compacted verification works)

WillemKauf replied:
Added

bharathv commented on tests/rptest/tests/datalake/datalake_verifier.py:
```diff
@@ -24,22 +24,24 @@
 
```
Compaction is expected to block until translation happens. What additional coverage does verification with a compacted log with a fully translated iceberg table provide?

WillemKauf replied:
>Compaction is expected to block until translation happens

>What additional coverage does verification with a compacted log with a fully translated iceberg table provide?

That the aforementioned expectation is true (i.e Iceberg table is fully translated, log is fully compacted).

Do you think there is other verification that should be added here?

bharathv replied:
> That the aforementioned expectation is true (i.e Iceberg table is fully translated, log is fully compacted).

Correct me if I'm wrong but the verifier in the current form can also succeed if the topic got translated from a compacted log (hypothetically if the code violated the max_collectible_offset invariant), no?

WillemKauf replied:
Yes, I suppose that is true.

In its current form, we handle the cases where the iceberg table has as much or more information than the log, as we assume that it was translated before compaction of the log took place/didn't take place, but this doesn't verify that the case you described didn't occur.

bharathv replied:
Thanks, so I was wondering if we should instead test enforcement of max_collectible_offset which is a more critical invariant and leave the verifier as it is today.

WillemKauf replied:
I agree that perhaps we haven't covered the case where the max_collectible_offset invariant is violated, but I do think that the existing changes to the verifier are helpful to at least ensure the case in which it ISN'T violated is correct

I'll try to think of ways to better cover the critical invariant in a follow up PR, if that works for you.

bharathv commented on tests/rptest/tests/datalake/compaction_test.py:
```diff
@@ -118,3 +120,126 @@ def test_translation_no_gaps(self, cloud_storage_type):
             self.do_test_no_gaps(dl)
+
+
+class CompactionTest(RedpandaTest):
```
One of the motivations for removing offset translation was to use iceberg enabled topics with read replicas/topic recovery, wondering if its worth adding an e2e test for it.

WillemKauf replied:
Would you like to see a test added in this PR, or as a follow up PR?

bharathv replied:
Follow up PR is also fine. (just want to ensure nothing else is broken, other than offset translation before we declare it as working).

andrwng replied:
FWIW read replicas or topic recovery don't appear to be addressed by this PR.

Read replica translators won't be able to perform offset translation on anything, and topic recovery will likely require changes to what revisions and topic overrides get passed to the coordinators

I don't believe either are in scope of this work (which is really just to unblock compaction IIUC)

bharathv replied:
Good point about topic recovery needing more work. I think @~bashtanov is doing some work here in the migrations land.

> Read replica translators won't be able to perform offset translation on anything,

whats the use case here with read replicas though? IIRC its about being able to (sql) query an iceberg enabled RRR topic, thats it? RRR topic itself cannot do any translation locally.

andrwng replied:
That's right. But as is, RRR topics won't be able to call the `to_log_offset()` that's done in the translation path. I think it's reasonable to have the translator skip the offset-translation and pass a nullopt to the STM as the log offset.

bharathv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -293,6 +295,16 @@ partition_translator::do_translate_once(retry_chain_node& parent_rcn) {
     }
     auto read_end_offset = max_offset.value();
     if (read_end_offset < read_begin_offset) {
+        // Detect that redpanda offset has changed from last translation, but no
+        // kafka offsets were found.
+        auto highest_translated_log_offset = get_translated_log_offset(
+          _partition->raft()->log(), read_end_offset);
+        auto reset_error = co_await _stm->reset_highest_translated_offset(
+          read_end_offset,
+          highest_translated_log_offset,
+          _term,
+          wait_timeout,
+          _as);
```
nit: Is this needed? the logic will reset_highest_translated_offset() after next successful translation anyway?
(probably avoids unnecessary batches in the log)

WillemKauf replied:
Imagine a partition which is no longer being produced to. The `max_collectible_offset()` will be limited by the translated log offset at time of last successful translation.

By also resetting here, we can do a better job of keeping the highest translated log offset (MCO) up to date as non-data batches are produced.

What do you think?

bharathv replied:
> The max_collectible_offset() will be limited by the translated log offset at time of last successful translation.

We cannot compact offset translation batches anyway? (even if we advance `max_collectible_offset`)

> What do you think?

My thinking was to avoid unnecessary log appends and offset translation batches if we can, no big deal to leave as is.

WillemKauf replied:
> We cannot compact offset translation batches anyway? (even if we advance max_collectible_offset)

Yes, but isn't this also important for `delete` retention enforcement?

bharathv replied:
it is, the only offset translator batches that can accumulate in a quiescent state are probably raft configuration batches that get appended on leadership changes or partition moves. Unless something went wrong there shouldn't be too many such batches, so delaying retention is probably not an issue either. 

I'm ok to keep it as is.

bharathv commented on src/v/datalake/translation/tests/state_machine_test.cc:
```diff
@@ -74,8 +75,11 @@ struct translator_stm_fixture : stm_raft_fixture<stm> {
+          log, update);
         co_return co_await stm->reset_highest_translated_offset(
-          update, stm->raft()->term(), 5s, as);
+          update, log_update, stm->raft()->term(), 5s, as);
```
nit: this change belongs to a different commit?

WillemKauf replied:
Moved commit

bharathv reviewed (commented):
sorry for the delay here.

WillemKauf committed:
`datalake`: add `translation/utils.cc`

And most importantly, add the function `highest_log_offset_below_next()`.
This function will be used to compute the appropriate highest translated
log offset for a given translated kafka offset while taking into account
translator batches.

This will allow us to be less pessimistic about the `max_collectible_offset`
returned by the `translation_stm` in the future.

WillemKauf commented:
Force push to:
* Rebase to `upstream/dev` and fix merge conflicts
* Rename `new_log_translated_offset` -> `new_translated_log_offset` in `state_machine.cc/.h`
* Add comment in `datalake_verifier.py` around `_expected_compacted_keys`
* Move changes for `state_machine_test.cc` to proper commit

WillemKauf commented:
Force push to:

* Rebase to `upstream/dev` and fix merge conflicts

andrwng commented on src/v/datalake/translation/state_machine.h:
```diff
@@ -39,19 +39,28 @@ class translation_stm final : public raft::persisted_stm<> {
+        }
     };
     kafka::offset _highest_translated_offset{};
+    model::offset _highest_translated_log_offset{};
```
What does it mean for this to be uninitialized? Have you considered making it `optional`? `std::nullopt` might feel like an intuitive value for a read replica, for instance

andrwng commented on src/v/datalake/translation/utils.h:
```diff
@@ -0,0 +1,37 @@
+// translator (Config) batch, TLO is the translated log offset, and NKO is the
+// next expected kafka record, we should expect TLO to be equal to the offset of
+// the last configuration batch before the next kafka record.
+model::offset
+get_translated_log_offset(ss::shared_ptr<storage::log> log, kafka::offset o);
```
Given it may be confusing to refer to "translation" here in multiple contexts (i'm actually not sure if you mean it as offset-translated or datalake-translated), it may be more self-descriptive if this were named `highest_log_offset_below(kafka::offset)`, where the translator would pass in `kafka::next_offset(max_translated_offset)`

andrwng replied:
Or consider the name `highest_log_offset_below_next()`

WillemKauf replied:
Being able to call this function in `translation_stm::max_collectible_offset()` like

`return get_translated_log_offset(_raft->log(), _highest_translated_offset);`

feels better than having to manipulate the passed offset outside the function before calling it each time.

If you feel strongly about this I can change the name- it is unfortunate that "translation" can mean two different things in this context but I hope the code comments are descriptive enough?

andrwng replied:
I feel strongly about the name -- we already have some many named offsets here and there, this one doesn't seem so pivotal that it needs a special name. My vote is for `highest_log_offset_below_next()`

andrwng commented on tests/rptest/tests/datalake/datalake_verifier.py:
```diff
@@ -77,6 +79,21 @@ def __init__(self, redpanda: RedpandaService, topic: str,
         self._last_checkpoint = {}
 
+        self._compacted = compacted
+        # When consuming from a compacted topic,
+        # there may be records in the Iceberg table
+        # that have since been compacted away in the log.
+        # We maintain a set of compacted keys during message
+        # verification. If the offset for a record read from
+        # the Iceberg table differs from the offset for the
+        # record read from the log, the record's key is added to
+        # the set. The key is removed from the set when a later
+        # record with the same key is seen in the log. Finally,
+        # after consuming, we assert that the size of this set is
+        # zero (otherwise, it would imply an anomaly between the
+        # Iceberg table and the log).
```
nit: this spacing's a bit funky?

WillemKauf replied:
In terms of line width?

andrwng replied:
Yeah, it seems inconsistent with itself

andrwng commented on src/v/datalake/translation/tests/translated_log_offset_test.cc:
```diff
@@ -0,0 +1,367 @@
+// an expected offset for a translated kafka offset.
+//
+// Returns true iff expected_offset == the highest translated log offset.
+bool check_translated_log_offset(
+  ss::shared_ptr<storage::log> log,
+  kafka::offset translated_offset,
+  model::offset expected_offset) {
+    auto translated_log_offset
+      = datalake::translation::get_translated_log_offset(
+        log, translated_offset);
+    return expected_offset == translated_log_offset;
+}
```
nit: this feels like quite a lot of testing for something that is ultimately just calling into the offset translator, and so this feels like we're just testing the offset translator. Wondering if we can instead write tests that check the max collectible offset? Plus, if you go the `highest_log_offset_below()` route, this also all changes

WillemKauf replied:
Yeah, ultimately this is more of an offset translation test, but I wanted to have a set of very illustrative unit tests that only tested this mechanism, and not the state machine as a whole.

andrwng commented on src/v/datalake/translation/utils.h:
```diff
@@ -0,0 +1,37 @@
+// has been produced to the log. This function will always return a value, and
+// its correctness depends on the validity of the input offset.
+//
+// For example, in the following situation:
+// Kaf offsets: [TO]  .   .    .     [NKO]
+// Log offsets: [K]  [C] [C] [C/TLO] [NKO]
+// where TO is the translated offset, K is the last kafka record, C is a
```
nit: this description imposes some expectation on what the input offset is. But in the implementation, it doesn't matter whether `o` has been translated or not, does it? Might be easier to grock this if we call `O` the "input offset" instead of calling `TO` the "translated offset"

andrwng commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -116,21 +138,23 @@ model::offset translation_stm::max_collectible_offset() {
     if (_highest_translated_offset == kafka::offset{}) {
         return model::offset{};
     }
-    return _raft->log()->to_log_offset(
-      kafka::offset_cast(_highest_translated_offset));
```
Curious, could this PR have been

```
return model::prev_offset(
    _raft->log()->to_log_offset(
    kafka::offset_cast(kafka::next_offset(_highest_translated_offset))));
```

without additional persistent state changes?

WillemKauf replied:
Yeah, it's starting to feel like it. I can't think of any other side effects/benefits to persisting both offsets.

It may be cleaner to leave the offset translation within the state_machine rather than storing two offsets everywhere, but I don't have a strong opinion one way or another. Let me know what you prefer

andrwng replied:
Yea, it'll be much more difficult to undo if we bake this into persisted storage. If the change works with just the change to the `max_collectible_offset()` function (which I'm not certain it will, but it does seem like it), let's only do that.

WillemKauf replied:
Removed all persistent state changes.

WillemKauf committed:
`datalake`: use `highest_log_offset_below_next()` in `translation_stm`

This will return a less restrictive value for
`translation_stm::max_collectible_offset()`.

WillemKauf commented:
Force push to:
* Remove persistence of `_highest_translated_log_offset` in `datalake/translation` structures
* Directly call into `get_translated_log_offset()` within `translation_stm::max_collectible_offset()`
* Add check for RRR in `translation_stm::max_collectible_offset()`

andrwng commented:
>Add check for RRR in translation_stm::max_collectible_offset()

Mind removing this from this PR and following up with a test in a separate PR? IMO this one here should be focused on unblocking compaction

WillemKauf commented:
Force push to:
* Rename `get_translated_log_offset()` -> `highest_log_offset_below_next()`
* Remove commit `Add check for RRR in translation_stm::max_collectible_offset()`
* Refactor code comment text width in `datalake_verifier.py`

WillemKauf commented:
Force push to:
* Rebase to `upstream/dev` to fix linter CI issues

WillemKauf commented:
Force push to: 
* Fix `bazel` build deps

bharathv reviewed (approved):
lgtm , cover letter needs to be updated to `highest_log_offset_below_next` ?

dotnwat commented on src/v/datalake/translation/utils.h:
```diff
@@ -0,0 +1,38 @@
+
+namespace datalake::translation {
+
+// Returns the equivalent log offset which can be considered translated by the
+// datalake subsystem, while taking into account translator batch types, for a
+// given kafka offset.
+//
+// Note that the provided kafka::offset o MUST be a valid offset, i.e one that
+// has been produced to the log. This function will always return a value, and
+// its correctness depends on the validity of the input offset.
+//
+// For example, in the following situation:
+// Kaf offsets: [O]   .   .    .     [NKO]
+// Log offsets: [K]  [C] [C] [C/TLO] [NKO]
+// where O is the input offset, K is the last kafka record, C is a translator
+// (Config) batch, TLO is the translated log offset, and NKO is the next
+// expected kafka record. We should expect TLO to be equal to the offset of the
+// last configuration batch before the next kafka record.
+model::offset highest_log_offset_below_next(
+  ss::shared_ptr<storage::log> log, kafka::offset o);
```
🥳
